sklearn/metrics/_regression.py

```diff
@@ -1197,3 +1201,235 @@ def d2_tweedie_score(y_true, y_pred, *, sample_weight=None, power=0):
     denominator = np.average(dev, weights=sample_weight)
 
     return 1 - numerator / denominator
+
+
+def d2_pinball_loss_score(
+    y_true, y_pred, *, sample_weight=None, alpha=0.5, multioutput="uniform_average"
+):
+    """
+    :math:`D^2` regression score function, \
+    fraction of pinball loss deviance explained.
+
+    Best possible score is 1.0 and it can be negative (because the model can be
+    arbitrarily worse). A model that always uses the empirical median of `y_true`
+    as constant prediction, disregarding the input features,
+    gets a :math:`D^2` score of 0.0.
+
+    Read more in the :ref:`User Guide <d2_score>`.
+
+    .. versionadded:: 1.1
+
+    Parameters
+    ----------
+    y_true : array-like of shape (n_samples,) or (n_samples, n_outputs)
+        Ground truth (correct) target values.
+
+    y_pred : array-like of shape (n_samples,) or (n_samples, n_outputs)
+        Estimated target values.
+
+    sample_weight : array-like of shape (n_samples,), optional
+        Sample weights.
+
+    alpha : float, default=0.5
+        Pinball loss quantile parameter, determines the slope of the pinball_loss.
+        Equivalent to `d2_absolute_error_score` when `alpha=0.5`.
+
+    multioutput : {'raw_values', 'uniform_average'} or array-like of shape \
+            (n_outputs,), default='uniform_average'
+        Defines aggregating of multiple output values.
+        Array-like value defines weights used to average scores.
+
+        'raw_values' :
+            Returns a full set of errors in case of multioutput input.
+
+        'uniform_average' :
+            Scores of all outputs are averaged with uniform weight.
+
+    Returns
+    -------
+    score : float or ndarray of floats
+        The :math:`D^2` score with a pinball loss deviance
+        or ndarray of scores if `multioutput='raw_values'`.
+
+    Notes
+    -----
+    This is not a symmetric function.
+
+    Like :math:`R^2`, :math:`D^2` score may be negative
+    (it need not actually be the square of a quantity D).
+
+    This metric is not well-defined for single samples and will return a NaN
+    value if n_samples is less than two.
+
+     References
+    ----------
+    .. [1] Eq. (3.11) of Hastie, Trevor J., Robert Tibshirani and Martin J.
+           Wainwright. "Statistical Learning with Sparsity: The Lasso and
+           Generalizations." (2015). https://trevorhastie.github.io
```

Comment:
```suggestion
```
Or it is actually mentioned in this reference?

Reply:
I thought this reference is quite general and can be a reference for the D2 family. The screenshot of the equation is attached below. It's challenging to find a good reference for the general D2 (or D2 with pinball loss), but I'm happy to add some other reference if you have one or just leave it empty if it's less confusing that way.
![image](https://user-images.githubusercontent.com/41272397/148874010-423b53c8-2015-4806-b355-6387c4647ac7.png)

Reply:
Thanks for confirming. Though we have a little problem. Technically, the pinball loss is no deviance. I'm not sure if we should just go with this (good) reference or if we should be precise and mention and reference the correct term "skill score".

Side note:
The more general term skill score stems from the forecast literature, `skill score = (S_m - S_ref) / (S_opt - S_ref)` where S is any scoring function / metric, S_M the one of the model `m`, `S_ref` the one of some reference model (a constant model in our case) and `S_opt` the one of a perfect model, ideally being zero as for the pinball loss.

Reply:
What definition do you use for deviance?
If we follow the definition on [wikipedia](https://en.wikipedia.org/wiki/Deviance_(statistics)), the unit pinball loss satisfies all the necessary conditions: `d(y, y) = 0` and `d(y, u) != 0` when `y != u`. The mean pinball loss we are using is equivalent to the total deviance from their definition up to a normalization constant (not a problem for D2 because it's canceled by the same constant in the denominator).

Reply:
Deviances are defined for the exponential (dispersion) family and can be derived from the log-likelihood. Therefore, pinball loss does not qualify.

Reply:
Can you please explain why pinball loss doesn't qualify?
If we follow the definition from [Peter X.-K. Song](https://link.springer.com/book/10.1007%2F978-0-387-71393-9)'s book (attached below). The unit pinball loss is deviance when alpha is in the open domain (0,1). Inserting the pinball loss unit deviance into equation (2.3) defines the member of the exponential dispersion family from which the pinball deviance is derived.
![image](https://user-images.githubusercontent.com/41272397/149646346-39cea06f-3e8c-4f0d-a616-7be3f154baf2.png)

Reply:
See https://stats.stackexchange.com/a/274926, i.e. exp(-deviance) is not a sufficient condition for the exponential family.

Reply:
Good point, the probability distribution from my previous comment (with unit pinball loss deviance) is a Laplace-like probability distribution (exactly Laplace for alpha=0.5) and it may not be part of the exponential dispersion family. Thanks for highlighting that! 
Luckily, the book does not limit the deviance definition to the exponential dispersion family (which is introduced in the introduction section 2.1), it instead defines it for the dispersion models family (section 2.2). It specifically clarifies that this family extends the euclidian distance of the normal distribution to a general discrepancy function d(y, u) and each family member is defined uniquely by d (equation 2.3). It also mentions that many of the exponential dispersion family members are members of this family as well.
![image](https://user-images.githubusercontent.com/41272397/149713847-b5b8e1e8-4ae1-4980-baae-bbd65dfb9549.png)

Reply:
Right, Jørgensen even mentions the absolute error. (Just keep in mind, that in general one needs to proof existence, e.g., the Tweedie deviance for 0<p<1 fails to exist. Let's stop here:wink:)